README
encdet - an encoding detection library

* About

encdet is an encoding detection library which is a C interface to
Mozilla Universal Charset Detector.
http://www.mozilla.org/projects/intl/ChardetInterface.htm

The original code is available at
http://mxr.mozilla.org/seamonkey/source/extensions/universalchardet/

The latest version is available at:
http://wiki.github.com/spiritloose/encdet

* Installation

./configure && make && make install

See INSTALL for more details.

* API

See encdet(3).

* License

Mozilla Public License Version 1.1.
Alternatively, the library may be used under the terms of either
the GNU General Public License Version 2 or later, or the GNU
Lesser General Public License 2.1 or later.

See COPYING and http://www.mozilla.org/MPL/ for more details.

* Author

Jiro Nishiguchi <jiro@cpan.org>



NEWS
encdet 0.2.0

* Merged upstream commits a3a6183d5a22 and 4dd0cf2ba86f.
  a3a6183d5a22: Bug 488426 - enable TIS-620 detector; r=smontagu
  4dd0cf2ba86f: Bug 493781 Fixing warning in universalchardet array
                subscript has type 'char'

encdet 0.1.0

* Initial release


test/test.sh
#!/bin/sh
# -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is encdet.
#
# The Initial Developer of the Original Code is
# Jiro Nishiguchi <jiro@cpan.org>
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****

warn() {
    echo $@ 1>&2
}

cd `dirname $0`
index=1
files=`find . -type f -name "*.txt" | sort`
tests=`echo "$files" | wc -l`
tests=`expr $tests + 0`
fail=0
echo "$index..$tests" # plan
for file in $files
do
    result=`./encdet $file`
    fileenc=`echo "$file" | sed -e 's/^\.\///' | sed -e 's/\.txt$//'`
    if [ "$result" != "$fileenc" ]; then
        echo "not ok $index - $fileenc"
        warn "#   Failed test '$fileenc'"
        warn "#   at $0 line $LINENO."
        warn "#          got: '$result'"
        warn "#     expected: '$fileenc'"
        fail=`expr $fail + 1`
    else
        echo "ok $index - $fileenc"
    fi
    index=`expr $index + 1`
done
if [ $fail != 0 ]; then
    warn "# Looks like you failed $fail tests of $tests."
    exit 1
fi
exit 0


test/encdet.c
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is encdet.
 *
 * The Initial Developer of the Original Code is
 * Jiro Nishiguchi <jiro@cpan.org>
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#include "encdet.h"
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

int
print_encoding(FILE *fp)
{
    char buf[BUFSIZ];
    const char *encoding;
    encdet_t det;
    int rc = 0;

    det = encdet_new(ENCDET_ALL);
    if (!det) {
        fprintf(stderr, "failed to encdet_create().\n");
        return rc;
    }
    while (fgets(buf, sizeof(buf), fp) != NULL) {
        rc = encdet_handle_data(det, buf, strlen(buf));
        if (rc) {
            fprintf(stderr, "failed to encdet_handle_data().\n");
            encdet_destroy(det);
            return rc;
        }
    }
    encdet_data_end(det);
    encoding = encdet_get_result(det);
    if (encoding) {
        puts(encoding);
    } else {
        puts("ASCII");
    }
    encdet_destroy(det);
    return 0;
}

int
main(int argc, char **argv)
{
    FILE *fp;
    int rc;
    if (argc == 1) {
        fp = stdin;
    } else if (argc == 2) {
        fp = fopen(argv[1], "r");
        if (fp == NULL) {
            perror(NULL);
            return EXIT_FAILURE;
        }
    } else {
        fprintf(stderr, "usage: encdet [FILE]\n");
        return EXIT_FAILURE;
    }
    rc = print_encoding(fp);
    if (argc == 2) fclose(fp);
    return rc ? EXIT_FAILURE : EXIT_SUCCESS;
}



src/prmem.h
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is the Netscape Portable Runtime (NSPR).
 *
 * The Initial Developer of the Original Code is
 * Netscape Communications Corporation.
 * Portions created by the Initial Developer are Copyright (C) 1998-2000
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *      Jiro Nishiguchi <jiro@cpan.org>
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#ifndef PRMEM_H
#define PRMEM_H

#include <cstdlib>

inline void* PR_Malloc(size_t len)
{
    return malloc(len);
}

#define PR_FREEIF(p) if (p) free(p)

#endif /* PRMEM_H */


src/encdet.h
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is encdet.
 *
 * The Initial Developer of the Original Code is
 * Jiro Nishiguchi <jiro@cpan.org>
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#ifndef ENCDET_H
#define ENCDET_H

#ifdef __cplusplus
extern "C" {
#if 0
} /* satisfy cc-mode */
#endif
#endif

#define ENCDET_CHINESE_SIMPLIFIED  0x01
#define ENCDET_CHINESE_TRADITIONAL 0x02
#define ENCDET_JAPANESE            0x04
#define ENCDET_KOREAN              0x08
#define ENCDET_NON_CJK             0x10
#define ENCDET_ALL                 0x1F
#define ENCDET_CHINESE (ENCDET_CHINESE_SIMPLIFIED | \
                        ENCDET_CHINESE_TRADITIONAL)
#define ENCDET_CJK (ENCDET_CHINESE_SIMPLIFIED | \
                    ENCDET_CHINESE_TRADITIONAL | \
                    ENCDET_JAPANESE | \
                    ENCDET_KOREAN)

typedef void* encdet_t;

encdet_t encdet_new(unsigned int langflags);

int encdet_handle_data(encdet_t det, const char *data, unsigned int len);

void encdet_data_end(encdet_t det);

void encdet_reset(encdet_t det);

const char *encdet_get_result(encdet_t det);

void encdet_destroy(encdet_t det);

const char *encdet_version(void);

#ifdef __cplusplus
#if 0
{ /* satisfy cc-mode */
#endif
} /* extern "C" { */
#endif

#endif /* ENCDET_H */


src/nscore.h
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is mozilla.org code.
 *
 * The Initial Developer of the Original Code is
 * Netscape Communications Corporation.
 * Portions created by the Initial Developer are Copyright (C) 1998
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *      Jiro Nishiguchi <jiro@cpan.org>
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either of the GNU General Public License Version 2 or later (the "GPL"),
 * or the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#ifndef NSCORE_H
#define NSCORE_H

typedef bool PRBool;
typedef unsigned char PRUint8;
typedef int PRInt32;
typedef unsigned int PRUint32;
typedef short PRInt16;
typedef unsigned short PRUint16;

#define PR_FALSE false
#define PR_TRUE true
#define nsnull 0

enum nsresult
{
    NS_OK,
    NS_ERROR_OUT_OF_MEMORY
};

#endif /* NSCORE_H */


src/encdet.cpp
/* -*- Mode: C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is encdet.
 *
 * The Initial Developer of the Original Code is
 * Jiro Nishiguchi <jiro@cpan.org>
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include "nscore.h"
#include "nsUniversalDetector.h"
#include "encdet.h"

class EncDet : public nsUniversalDetector
{
public:
    EncDet(PRUint32 aLanguageFilter) : nsUniversalDetector(aLanguageFilter) {}

    virtual ~EncDet() {}

    virtual void Report(const char *charset)
    {
        mDetectedCharset = charset;
    }

    virtual void Reset()
    {
        this->nsUniversalDetector::Reset();
    }

    const char* GetResult() const
    {
        return mDetectedCharset;
    }
};

#define CAST(x) reinterpret_cast<EncDet*>(x)

encdet_t
encdet_new(unsigned int langflags)
{
    return reinterpret_cast<encdet_t>(new EncDet(langflags));
}

int
encdet_handle_data(encdet_t det, const char *data, unsigned int len)
{
    if (!det) return -1;
    nsresult ret = CAST(det)->HandleData(data, (PRUint32)len);
    return ret == NS_OK ? 0 : -1;
}

void
encdet_data_end(encdet_t det)
{
    if (det) CAST(det)->DataEnd();
}

void encdet_reset(encdet_t det)
{
    if (det) CAST(det)->Reset();
}

const char *
encdet_get_result(encdet_t det)
{
    if (!det) return 0;
    return CAST(det)->GetResult();
}

void
encdet_destroy(encdet_t det)
{
    if (det) delete CAST(det);
}

const char *
encdet_version()
{
    return PACKAGE_VERSION;
}

